ios/Classes/codec/AblyFlutterReader.h

```diff
@@ -0,0 +1,28 @@
+@import Foundation;
+@import Flutter;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface AblyFlutterReader : FlutterStandardReader
+@end
+
+NS_ASSUME_NONNULL_END
+
+
+typedef NS_ENUM(UInt8, _Value) {
+    //Ably flutter plugin protocol message
+    _valueAblyMessage = 128,
+    
+    //Other ably objects
+    _valueClientOptions = 129,
```

Comment:
More craziness...  Sorry. I just cannot warm to this.

Reply:
@mattheworiordan while we can solve the ordering issue by using a map/dict, there is nothing we can do here. Flutter's [implementation to encode and decode default supported types](https://flutter.dev/docs/development/platform-integration/platform-channels?tab=ios-channel-objective-c-tab#codec) also follows [the same approach](https://github.com/flutter/flutter/blob/f139b11009/packages/flutter/lib/src/services/message_codecs.dart#L265), so I would say we should get used to this :)

If you have any inputs/alternate suggestions, we can explore that.

Reply:
@tiholic perhaps we could get rid of our custom codec entirely and **always** use a `Map` for objects?

You'll then need a set of serialisers in each language that are capable of encoding and/or decoding the classes/structs that need to go across the native boundary.  Where there are trees of nested structures to be encoded then clearly the encoding serializers also need to nest accordingly to match. Same for the decoding serializers too.

In terms of map keys, those can then be localised to the decoder implementation and can all be of type `String` so that their representations conform exactly to the names in [the spec](https://docs.ably.io/client-lib-development-guide/features/). Essentially, a JSON-like approach - hence a good comparable is the functionality of Apple's [NSJSONSerialization](https://developer.apple.com/documentation/foundation/nsjsonserialization) in terms of how that implementation has been engineered only to understand a focussed subset of the [Foundation](https://developer.apple.com/documentation/foundation?language=objc) data types.

Clearly these encoding and decoding serialisers are no longer coupled to the codec. We leave that in it's Flutter-native form as the 'transport layer'. It does what it does and then we handle our own protocol in our own layer above that.

As a final thought, perhaps worth keeping [YAGNI](https://martinfowler.com/bliki/Yagni.html) in mind so not writing a decoder or encoder for a particular platform if you don't need one. A serializer should be capable of being either a decoder, or encoder, or both, as required.

Reply:
Created an issue for further discussions on this.
I will explore on your inputs and share my feedback @ https://github.com/ably/ably-flutter/issues/12